lib/pyfrc/physics/visionsim.py

```diff
@@ -0,0 +1,189 @@
+
+import collections
+import math
+
+inf = float('inf')
+twopi = math.pi*2.0
+
+def _in_angle_interval(x, a, b):
+    return (x - a) % twopi <= (b - a) % twopi
+
+class VisionSimTarget:
+    def __init__(self, x, y, view_angle_start, view_angle_end):
+        '''
+            :param x: Target x position
+            :param y: Target y position
+        
+            :param view_angle_start:
+            :param view_angle_end: clockwise from start angle
+            
+            View angle is defined in degrees from 0 to 360, with 0 = east,
+            increasing clockwise. So, if the robot could only see
+            the target from the south east side, you would use a view angle
+            of start=0, end=90.
+        '''
+        self.x = x
+        self.y = y
+        
+        self.view_angle_start = math.radians(view_angle_start)
+        self.view_angle_end = math.radians(view_angle_end)
+
+    def compute(self, now, x, y, angle):
+        
+        # incoming angle must be normalized to -180,180 (done in VisionSim)
+        
+        # note: the position tracking of the robot has Y incrementing
+        # positively downwards, presumably because it's convenient for
+        # drawing stuff. Not what I had expected.
+        
+        # compute target distance
+        dx = self.x - x
+        dy = self.y - y
+        
+        distance = math.hypot(dx, dy)
+        
+        # at the right distance?
+        if distance < self.view_dst_start or distance > self.view_dst_end:
+            return
+            
+        # determine the absolute angle of the target relative to the robot
+        target_angle = math.atan2(dy, dx)
+        
+        # in the view range?
+        if not _in_angle_interval(target_angle + math.pi,
+                                  self.view_angle_start,
+                                  self.view_angle_end):
+            return
+        
+        # is the robot facing the target?
+        a = angle - self.fov2
+        b = angle + self.fov2
+        
+        if _in_angle_interval(target_angle + math.pi, a + math.pi, b + math.pi):
+            offset = math.degrees((((target_angle - angle) + math.pi) % (math.pi*2)) - math.pi)
```

Comment:
Probably a good idea to use `math.atan2(math.sin(angle), math.cos(angle))` rather than float modulo.